### DISCUSSION TITLE
Visual grouping of nodes

### DISCUSSION THREAD

**Discussion by moklick** [xyflow/xyflow#1022] [New Features]
Add a `group` option for nodes that draws a box around all nodes that belong to that group.

**Comment by zulus23** [comment]
It's good idea  have to opportunity  grouping nodes. It's be better have to opportunity roll up and unfold group yet

**Reply by shushuitie2017** [reply]
It's good idea！

**Comment by shushuitie2017** [comment]
It's good idea！

**Comment by pnkflydgr** [comment]
@moklick, I am interested in a parent/child relationship for node boundaries. 

A parent node could have a port(s), but also have child nodes that port(s) and also have child nodes.  My use case would need to be 3 levels deep.  Parent -> children -> children -> children.  The children could not be moved outside of their parents.

I need the children ports to be attachable from nodes outside of the parent node.

Does this use case fit here or should I create a separate issue?

**Reply by moklick** [reply]
I think you are speaking about nested flows https://github.com/wbkd/react-flow/discussions/1024

**Comment by mikest** [comment]
I have this exact use case. Right now I'm hacking it with a "portless" node that gets updated after edits, but at some point I'll need to make it real. I would be up for collaborating.

![image](https://user-images.githubusercontent.com/742517/114435003-841ff600-9b78-11eb-81c8-fee97986da29.png)

**Reply by mikest** [reply]
For our system, the group could either be thought of us as "sub nodes" which would move this to a different feature, or as "a group of nodes with boundary IO nodes." which is probably an abuse of the grouping concept.

**Reply by pdubey84** [reply]
@mikest , did you already make this work? Anyway you can share the code gist/codesandbox for this?

**Reply by mikest** [reply]
@pdubey84 we did not make this work in any portable way, unfortunately :/

**Comment by paulodeon** [comment]
I also need the children ports to be attachable from nodes outside of the parent node.

**Comment by IOR88** [comment]
Hi @moklick I think the grouping feature will add a lot of value to this project, I am very interested in this feature, could You tell how I could help in development of this feature ? Do You have some release date planned ? Best, Igor.

**Comment by CamboimGabriel** [comment]
I think that would be a nice addition too!

**Comment by jackson-daniel** [comment]
Also interested in this feature. Would be great if it is added to the roadmap.

**Comment by natakina** [comment]
It would be great! @moklick Are you planning to do it soon?

**Comment by xpluscal** [comment]
I feel that this could also just be something like "Nesting Nodes".

Aka how this could work:
groupNodes are also "just" nodes, but other nodes in the list get an attribute "parent" with supplying the id of the groupNode.

**Comment by truong-hua** [comment]
That's definitely perfect idea especially when the number of nodes in a diagram is large enough. Like other design tool (Adobe Illustrator, Photoshop), it is better if we can support group isolation so that user can focus to edit just a small set of items.

It would make react-flow becomes an enterprise ready flow design framework.

**Comment by datoslabs** [comment]
I have some experimental code that does the following:
![untitled](https://user-images.githubusercontent.com/39677444/137972107-3e254941-6d7c-496b-9b0a-3b57b325a463.gif)

The code is rather messy at the moment mainly because I was trying out to see if it can be done without changing the core library.  Seeing as there seems to be a lot of interest, if there are a few ppl willing to collaborate, I am happy to spend some time to clean it up a bit and work/share with the team to implement this.  A lot of my code can be simplified if we were to change/add-to the core react-flow library (especially the ability to render children nodes as nested `divs` and adding to the store functions).

**Reply by truong-hua** [reply]
It's great UX, is it possible to package it as a plugin for current version?

**Reply by moklick** [reply]
wow. This looks awesome :o

**Reply by pdubey84** [reply]
Looks superb. Would you be ok to put up a PR for this or a separate branch for folks to see and collaborate

**Reply by datoslabs** [reply]
I think a separate branch would be easier for collaboration.  I would be happy to clean up and push my code; however, I feel a lot of the code I wrote (such as handling sizing/resizing of parent node which involves messy calculations of children node boundaries and positions) can much simplified using nested `divs` (ie render children nodes as nested `divs` under the parent node `div`) thus delegating the layout/rendering to CSS.  This will require changes to `NodeRenderer` and most likely other core libraries/reducers.

**Reply by jackson-daniel** [reply]
+1 would be great if added to a branch!

**Reply by ghachemEya** [reply]
Hello, i would like to know more about this code, i have a similar issue with parent and child node with React flow, Thank you !

**Reply by moklick** [reply]
Hey @ghachemEya In a few weeks we will publish v10. That version comes with grouping and sub flows. You can find the details in this PR https://github.com/wbkd/react-flow/pull/1555

**Reply by Samadmehmood** [reply]
@datoslabs can you share link to for the code example which makes such nested flows? i am trying react flow but the issue is how can i handle a drop when one node is droped on another parent node, i cannot achiveve that functionality? can u help?
Thanks!

**Reply by datoslabs** [reply]
@Samadmehmood, have you tried V10?  I am still on V9.x and have not upgraded to V10 bc I was working on another part of the project atm.

**Reply by Samadmehmood** [reply]
i am trying v10 rightnow but the issue is tat i cannot get to know whenever a node is dragged inside another node. how can i achieve that. to make it work like the container thing you have above.

**Comment by CamboimGabriel** [comment]
I implemented by myself too

![image](https://user-images.githubusercontent.com/36650520/138001740-f77bd5e7-11e4-48a0-bdff-fd3d83ac4825.png)

**Comment by moklick** [comment]
We are currently working on this feature. Which API do you prefer? We have a little voting here: https://github.com/wbkd/react-flow/discussions/1024#discussioncomment-1512393

**Reply by iti92** [reply]
Hi,
this is a wonderful news!
can you please provide us a date of first realease of grouping?

**Reply by moklick** [reply]
You can find the details in this PR #1555 (hopefully end of February).

**Comment by moklick** [comment]
This feature is available since v10 https://reactflow.dev/docs/guides/sub-flows/
